### PR TITLE
Improving slicing for nn.ModuleList to support arbitrary step size

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58361 More robust slicing logic for nn.ModuleList**

This PR makes the slicing logic for nn.ModuleList more robust. Previously, it was only supported for step size of 1 . Since we already implemented more robust tuple slicing before where we handle arbitrary step size, we made sure that the slicing logic for nn.ModuleList follows the same pattern as general tuple slicing. 
Fixes: https://github.com/pytorch/pytorch/issues/58360

Differential Revision: [D28464855](https://our.internmc.facebook.com/intern/diff/D28464855)

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 1]
```diff
@@ -4208,20 +4208,29 @@ struct to_ir {
...
         }
-        // pushing step_size to match the tuple_args
-        tuple_args.emplace_back(c10::nullopt);
 
+        if (slice.step().present()) {
+          auto step =
+              NamedValue(val_range, "step", emitExpr(Expr(slice.step().get())));
+          tuple_args.emplace_back(step);
+        } else {
+          tuple_args.emplace_back(c10::nullopt);
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/ir_emitter.cpp:4232]: Is this branch tested?

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 1 above)
Comment (tugsbayasgalan) [reply torch/csrc/jit/frontend/ir_emitter.cpp:4232]: done

File: test/jit/test_slice.py [hunk 2]
```diff
@@ -136,3 +136,25 @@ def tuple_slice(a):
...
+
+        scripted_foo = torch.jit.script(Foo())
+        result1, result2 = scripted_foo()
+
+        self.assertEqual(len(result1), 2)
+        self.assertEqual(str(result1[0]._type()), "__torch__.torch.nn.modules.rnn.GRU")
+        self.assertEqual(str(result1[1]._type()), "__torch__.torch.nn.modules.linear.Linear")
+
+        self.assertEqual(len(result2), 1)
+        self.assertEqual(str(result2[0]._type()), "__torch__.torch.nn.modules.activation.ReLU")
```
Comment (gmagogsfm) [inline test/jit/test_slice.py:160]: Type string comparison is fragile, please use a more robust method to verify test results.

One potential solution is to create a dummy module class that has a string-typed attribute. Then you can create 3 instances that each has a different value for that attribute. This is both lighter-weight and more robust than current implementation.

File: test/jit/test_slice.py [hunk 3]
```diff
@@ -136,3 +136,25 @@ def tuple_slice(a):
...
+        class Foo(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                m1 = torch.nn.Linear(1, 2)
+                m2 = torch.nn.ReLU()
+                m3 = torch.nn.GRU(1, 2)
+                self.test = torch.nn.ModuleList([m1, m2, m3])
+
+            def forward(self):
+                return self.test[::-2], self.test[1:2:]
```
Comment (gmagogsfm) [inline test/jit/test_slice.py:150]: These are not great test data.

For example, `[1:2:]` is bad for testing step size, any `step` size would result in same slicing behavior.

```
>>> a = [1, 2, 3]
>>> a[1:2:3]
[2]
>>> a[1:2:2]
[2]
>>> a[1:2:1]
[2]
```

**Review by gmagogsfm — CHANGES_REQUESTED** [review]

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

File: test/jit/test_slice.py (hunk 2 above)
Comment (tugsbayasgalan) [reply test/jit/test_slice.py]: done

File: test/jit/test_slice.py (hunk 3 above)
Comment (tugsbayasgalan) [reply test/jit/test_slice.py]: done

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28464855).

**Review by gmagogsfm — APPROVED** [review]

**Pull request closed without merging**